src/Types/TxConstraints.purs

```diff
@@ -246,6 +249,15 @@ mustPayToScript vh dt vl =
   singleton (MustPayToScript vh dt vl)
     <> singleton (MustIncludeDatum dt)
 
+mustPayToScriptInlineDatum
+  :: forall (i :: Type) (o :: Type)
+   . ValidatorHash
+  -> Datum
+  -> Value
+  -> TxConstraints i o
+mustPayToScriptInlineDatum vh dt vl =
```

Comment:
I know we want to be backwards-compatible, but I think we should just update `mustPayToScript` to take an `OutputDatum` instead. It's a breaking change, but I think upgrading won't be too bad

Reply:
Unfortunately we can't use `OutputDatum`, as the corresponding constructor for the normal `mustPayToScript` only holds the hash, not the datum.

Reply:
You mean `MustPayToScript`? AFAIK it takes a `Datum` argument and then hashes it

Reply:
No I mean the following DT:
https://github.com/Plutonomicon/cardano-transaction-lib/blob/4f84fceddbe9e7aecf319e0ceed0e9313ca1813f/src/Types/OutputDatum.purs#L30
`OutputDatumHash` is what corresponds to 'normal' datum providing, like in `mustPayToScript`. We need at least the datum, not just the hash

Reply:
Sorry, I still don't understand what you mean. If we change `mustPayToScript` to take an `OutputDatum` instead of a `Datum`, we can just add it to the `TransactionOutput` directly (perhaps checking to ensure that `NoOutputDatum` wasn't provided). If a user provides an `OutputDatumHash`, then there's no inline datum and it works like the current `mustPayToScript` implementation

Reply:
The current implementation of `mustPayToScript` adds both `MustPayToScript` and `MustIncludeDatum` constraints. While the former only requires the hash, the latter requires the full datum. Hence the full datum is required whether we're using inline datum or not, and `OutputDatumHash` only holds the hash

Reply:
Ah right, of course. I suppose having two constraints will work, although I would like to find a way to combine them at some point. We could add an additional arguments (maybe a sum type isomorphic to `Boolean`) to indicate whether or not the datum should be inline

Reply:
I've added a sum type, it's kinda ugly in practice though, at least in our examples with the qualified constraint import